### PR TITLE
Rav-1273 : Handling for NVL and NVL2 function

### PR DESCRIPTION
**Pull request by vaibhavjain-dm**
Rav-1273 : Handling for NVL and NVL2 function

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/sql/SqlKind.java [hunk 1]
```diff
@@ -488,6 +488,11 @@
    */
   NVL,
 
+  /**
+   * The "IFNULL" function (BigQuery).
+   */
+  IFNULL,
```
[the remark this answers is not in the document]
Comment (vaibhavjain-dm) [reply core/src/main/java/org/apache/calcite/sql/SqlKind.java]: Done

File: core/src/main/java/org/apache/calcite/rex/RexSimplify.java [hunk 2]
```diff
@@ -2316,6 +2318,21 @@ private RexNode simplifyIf(RexCall e, RexUnknownAs unknownAs) {
     }
     return e;
   }
+
+  private RexNode simplifyNvl(RexCall e, RexUnknownAs unknownAs) {
+    List<RexNode> operands = e.getOperands();
+    List<RexNode> resultRexNode = new ArrayList<>();
```
Comment (soma-mondal) [inline core/src/main/java/org/apache/calcite/rex/RexSimplify.java]: Please rename resultRexNode

File: core/src/main/java/org/apache/calcite/rex/RexSimplify.java [hunk 3]
```diff
@@ -2316,6 +2318,21 @@ private RexNode simplifyIf(RexCall e, RexUnknownAs unknownAs) {
...
+    for (RexNode operand : operands) {
+      resultRexNode.add(simplify(operand, unknownAs));
+    }
+    RexNode rexNode = simplifyIsNotNull(resultRexNode.get(0));
+    if (rexNode != null && rexNode.isAlwaysTrue()) {
+      return resultRexNode.get(0);
+    } else if (rexNode != null && rexNode.isAlwaysFalse()) {
+      return resultRexNode.get(1);
+    }
+    return e;
```
Comment (soma-mondal) [inline core/src/main/java/org/apache/calcite/rex/RexSimplify.java]: Kindly change as dscussed

File: core/src/main/java/org/apache/calcite/rex/RexSimplify.java [hunk 4]
```diff
@@ -2316,6 +2318,19 @@ private RexNode simplifyIf(RexCall e, RexUnknownAs unknownAs) {
     }
     return e;
   }
+
+  private RexNode simplifyNvl(RexCall e) {
```
Comment (sachinrohaj-29) [inline core/src/main/java/org/apache/calcite/rex/RexSimplify.java]: Add java doc on top of this function.

File: core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java [hunk 5]
```diff
@@ -4960,6 +4961,28 @@ public void testIfMethodArgument() {
     assertThat(toSql(root, DatabaseProduct.HIVE.getDialect()), isLinux(expectedHive));
   }
 
+  @Test public void testNvlFunctionRelToSql() {
```
Comment (sachinrohaj-29) [inline core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java]: remove this test from here and add it in sql functions.

File: core/src/test/java/org/apache/calcite/sql/test/SqlOperatorBaseTest.java [hunk 6]
```diff
@@ -352,6 +352,28 @@ protected SqlTester tester(SqlLibrary library) {
                 .with("fun", library.name()));
   }
 
+  protected SqlTester hiveTester() {
```
Comment (sachinrohaj-29) [inline core/src/test/java/org/apache/calcite/sql/test/SqlOperatorBaseTest.java]: missing test for BQ.

File: core/src/test/java/org/apache/calcite/test/JdbcTest.java [hunk 7]
```diff
@@ -6237,6 +6237,22 @@ public void testGetDate() throws Exception {
             "C=250",
             "C=500");
 
+    CalciteAssert.that(CalciteAssert.Config.REGULAR)
```
Comment (sachinrohaj-29) [inline core/src/test/java/org/apache/calcite/test/JdbcTest.java]: missing test for BQ.

File: core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java (hunk 5 above)
Comment (vaibhavjain-dm) [reply core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java]: Done

File: core/src/test/java/org/apache/calcite/sql/test/SqlOperatorBaseTest.java (hunk 6 above)
Comment (vaibhavjain-dm) [reply core/src/test/java/org/apache/calcite/sql/test/SqlOperatorBaseTest.java]: Done.

File: core/src/test/java/org/apache/calcite/test/JdbcTest.java (hunk 7 above)
Comment (vaibhavjain-dm) [reply core/src/test/java/org/apache/calcite/test/JdbcTest.java]: Done

File: core/src/test/java/org/apache/calcite/test/JdbcTest.java [hunk 8]
```diff
@@ -6243,6 +6243,34 @@ public void testGetDate() throws Exception {
...
+  @Test public void testNvl() {
+    CalciteAssert.that(CalciteAssert.Config.REGULAR)
+            .with(CalciteConnectionProperty.FUN, "hive")
+            .query("select nvl(\"commission\", -99) as c from \"hr\".\"emps\"")
+            .returnsUnordered("C=-99",
+                    "C=1000",
+                    "C=250",
+                    "C=500");
+
+    CalciteAssert.that(CalciteAssert.Config.REGULAR)
```
Comment (sachinrohaj-29) [inline core/src/test/java/org/apache/calcite/test/JdbcTest.java]: does it not need @Test annotation?

Comment (vaibhavjain-dm) [reply core/src/test/java/org/apache/calcite/test/JdbcTest.java]: @test is already added.

**Pull request merged**